### PR TITLE
Cleanup dockerize

### PR DESCRIPTION
**Pull request by bennahugo**
Easier to work through this via GH. Here is the current compilation log with the problem (I had to add a few things just even for this to pick up e.g. all the include paths (esp. compiling that test case).

[compilation.err.log](https://github.com/twillis449/ALBUS_ionosphere/files/9619740/compilation.err.log)

### CODE REVIEW

**Comment by bennahugo** [comment]
@twillis449 please see above along with my changes needed to get this far.

File: C++/mim/test/PIMrunner/ionosphere_iri.h [hunk 1]
```diff
@@ -32,7 +32,7 @@ namespace MIM_PIM {
 
 
 
-extern const char IRI_DATA_PATH_DEFAULT[] MIMSTR(INSTALLDIR"/libdata/IRI/");    
+extern const char IRI_DATA_PATH_DEFAULT[] MIMSTR(INSTALLDIR##"/libdata/IRI/");
```
Comment (bennahugo) [inline C++/mim/test/PIMrunner/ionosphere_iri.h:35]: @twillis449 I had to use concatenate here to make this compile ??? am I doing something stupid?

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+#### Ubuntu 18.04 ships Python 3.6 LTS not 3.8 as it is defined in the build system
+RUN sed -i '19s/.*/export CURRENT_PYTHON = python3.6/' $ALBUSPATH/Makefile
+RUN sed -i '22s/.*/export PYTHONINCLUDEDIR = \/usr\/include\/python3.6/' $ALBUSPATH/Makefile
+#### Build with specified GNU toolchain
+RUN sed -i '39s/.*/export CC = gcc-'"${GNUCOMPILER}"'/' $ALBUSPATH/Makefile
+RUN sed -i '40s/.*/export F77 = gfortran-'"${GNUCOMPILER}"' --std=legacy/' $ALBUSPATH/Makefile
+RUN sed -i '43s/.*/export F77_RECL_UNIT = bytes/' $ALBUSPATH/Makefile
+RUN sed -i '45s/.*/export C++ = g++-'"${GNUCOMPILER}"'/' $ALBUSPATH/Makefile
+RUN sed -i '47s/.*/export CPP = cpp-'"${GNUCOMPILER}"' -P/' $ALBUSPATH/Makefile
+RUN sed -i '10s/.*/CFLAGS += -I$(PYTHONINCLUDEDIR) -I$(INSTALLDIR)\/include -DINSTALLDIR=''$(INSTALLDIR)''/' $ALBUSPATH/C++/mim/test/PIMrunner/Makefile
```
Comment (bennahugo) [inline Dockerfile:97]: Have to pass -DINSTALLDIR to get the variable passed through gcc. Not sure how this worked previously

File: C++/mim/test/PIMrunner/ionosphere_iri.h (hunk 1 above)
Comment (bennahugo) [reply C++/mim/test/PIMrunner/ionosphere_iri.h]: nvm

**Comment by bennahugo** [comment]
Got a bit further now running into this...
```
 Processing the orthogonal polynomial coefficients file ...
At line 202 of file usutounf.f (unit = 1, file = 'fshsmhde.nop')
Fortran runtime error: Bad value during floating point read

Error termination. Backtrace:
#0  0x7f96ff8ff32a
#1  0x7f96ff8ffed5
#2  0x7f96ff90069d
#3  0x7f96ffa74d6d
#4  0x7f96ffa780a8
#5  0x7f96ffa7972c
#6  0x55ba3438f4d6
#7  0x55ba3438f881
#8  0x7f96ff176c86
#9  0x55ba34387dc9
#10  0xffffffffffffffff
make[3]: *** [install] Error 2
```

deep inside Fortran land with GNU 7
going to try GNU 6 -- I believe that is shipped with Ubuntu 18.04

**Comment by bennahugo** [comment]
hmm it appears this runs on GNU 6.... however I'm not sure whether this is actually a deterministic build error

**Comment by twillis449** [comment]
Amongst other things its really annoying that James Anderson tossed everything including a dog's breakfast into ALBUS. The  International Reference Ionosphere (IRI) stuff which you have had issues with is not even used anywhere for data reduction in ALBUS. Also He has all this VEX and SOFA software which can probabably be gotten rid of eventually. Unfortunately if one comments out the IRI section in the top level Makefile then other sections fail because of 'missing' include files. Ugh!

**Comment by twillis449** [comment]
As I recall libraries should get copied into ~/albus/lib as soon as they are generated.

**Comment by bennahugo** [comment]
Of course it will help considerably if I could spell ```LD_LIBRARY_PATH``` properly..... inbound just some last cleanup

**Comment by bennahugo** [comment]
Oh hell yea!!
```
chmod -R ugo+rX /optsoft/ALBUS/share
Removing intermediate container c52a4c002f4f
 ---> 23369f3ee2ea
Step 59/59 : RUN python -c "import AlbusIonosphere" && echo "Crack the bubbly - this hog is airborne!!!"
 ---> Running in 4780d8d610a1
Crack the bubbly - this hog is airborne!!!
Removing intermediate container 4780d8d610a1
 ---> 42b7da9f5a20
Successfully built 42b7da9f5a20
Successfully tagged albus:latest

```

**Comment by bennahugo** [comment]
Need to move that one last apt install up to the master list at the top, but that can really wait till tomorrow - I need to leave the office before security throws me out...

Here is a tutorial on how to get docker running on Ubuntu 18.04 @twillis449 
https://www.digitalocean.com/community/tutorials/how-to-install-and-use-docker-on-ubuntu-18-04

Take careful note to add the ```docker``` group to your user (and log out and back in as per usual) otherwise all docker commands to the daemon requires sudo. Be careful with docker though ... it is only intended as development helper -- don't use this on a user that you have publically exposed in any way shape or form (e.g. a Jenkins user with a public-facing login --- @o-smirnov can tell you all about root-escalations).

After that navigate to the root directory of the checked out branch (containing the Dockerfile)
Then run
```
docker build -t albus .
```

We can work on cleaning up and mounting scripts tomorrow

**Comment by bennahugo** [comment]
@twillis449

Ok cleaned up dockerfile and added a blurp with help string, build as
```
$ docker build -t albus .
```
should yield
```
Step 58/61 : RUN python -c "import AlbusIonosphere" && echo "Crack the bubbly - this hog is airborne!!!"
 ---> Running in f7d4ee8c14ac
Crack the bubbly - this hog is airborne!!!
Removing intermediate container f7d4ee8c14ac
 ---> 690482e6b95b
Step 59/61 : ENTRYPOINT [ "/usr/bin/python3.6" ]
 ---> Running in 07da21da7f07
Removing intermediate container 07da21da7f07
 ---> a226fa733f51
Step 60/61 : ENV HELPSTRING "docker run -v <absolute path to gfzrnx>:/optsoft/bin/gfzrnx ""-v <absolute path to your waterhole with scripts>:/albus_waterhole ""--workdir /albus_waterhole ""--rm ""--user $(id -u <your user>):$(id -g <your user>) ""albus:latest <path to script mounted inside the waterhole>"
 ---> Running in 43252b0c65e9
Removing intermediate container 43252b0c65e9
 ---> a78143cd9dad
Step 61/61 : CMD [ "-c", "import AlbusIonosphere; import os; import pkg_resources; print('ALBUS\\n====='); print(AlbusIonosphere.__doc__); version = pkg_resources.require('AlbusIonosphere')[0].version; print('Version {}'.format(version)); print('Usage: ' + os.environ['HELPSTRING'])" ]
 ---> Running in 070a2599ef86
Removing intermediate container 070a2599ef86
 ---> a736fa293499
Successfully built a736fa293499
Successfully tagged albus:latest
```
this excludes including gfzrnx which you will have to mount in manually when running. You can run and get the blurp as 
```
$ docker run -v /home/hugo/workspace/gfzrnx_2.0-8219_lx64:/optsoft/bin/gfzrnx -v acceptance_tests:/albus_waterhole --workdir /albus_waterhole --rm albus:latest

ALBUS
=====
Interface to c++ ionosphere stuff.
Version 1.0
Usage: docker run -v <absolute path to gfzrnx>:/optsoft/bin/gfzrnx -v <absolute path to your waterhole with scripts>:/albus_waterhole --workdir /albus_waterhole --rm --user $(id -u <your user>):$(id -g <your user>) albus:latest <path to script mounted inside the waterhole>
```

I've defined a special folder inside the container called /albus_waterhole to mount in scripts and get output (under the uid and gid you specify to docker run) to make workflows uniform

We can run your basic acceptance test like so (or any other script for that matter without the -m nose)
```
$docker run -v /home/hugo/workspace/gfzrnx_2.0-8219_lx64:/optsoft/bin/gfzrnx -v /home/hugo/workspace/ALBUS_ionosphere/acceptance_tests:/albus_waterhole --workdir /albus_waterhole --rm --user $(id -u hugo):$(id -g hugo) albus:latest -m nose /albus_waterhole

....
number of valid receivers 1
Starting total bias correction step 18
number of valid receivers 1
Starting total bias correction step 19
number of valid receivers 1
Starting total bias correction step 20
number of valid receivers 1
Starting total bias correction step 21
number of valid receivers 1
Got 1 receivers close to rec 0
Total bias correction finished
Wed Sep 28 17:45:38 UTC 2022
.
----------------------------------------------------------------------
Ran 1 test in 86.666s

OK
```

However I cannot get ```Sutherland``` and ```VLA``` to work

**Comment by bennahugo** [comment]
I wonder if the last errors are not LST issues... I will  play with the rise and sets

**Comment by bennahugo** [comment]
Hmm no albus good stations are empty.... I'm being stupid will fix

**Comment by twillis449** [comment]
The UT times you gave me should produce output for all 3 of those sites, assuming that you get get the Sutherland Rinex 3 file translated. The VLA data is there but would be pretty well useless as the elevation is very low,

**Comment by bennahugo** [comment]
Lets see if Jenkins calls... -- done with a cron job so lets give it 5

**Comment by bennahugo** [comment]
Alright @twillis449 I just need you to add a user called ```ratt-priv-ci``` to the contributors with write access and we should be good to go (just allows it to set status indicators and reports)

**Comment by twillis449** [comment]
I think I got the contributor thing to work. ratt-priv-ci  should get an invitation.

**Comment by bennahugo** [comment]
ok to test

**Comment by bennahugo** [comment]
@twillis449 yup received and accepted. Sorry there is likely going to be some spam commits while I configure the test -- it is always trial by error

**Comment by bennahugo** [comment]
Alright @twillis449 can you review to see if you are happy with this. It at least set a basis for me to work on changes without worrying about breakage - if there are other tests you can think we should add then they can all be placed in the acceptance tests with the same format -- I presume we are only testing a small portion of the entire codebase with this 1 test case

**Comment by twillis449** [comment]
I'm not sure what I'm supposed to do here for review etc.  If  you are getting data etc you have probably tested ~85 %  of the useful code base. As I've mentioned before there's a lot of stuff that can be thrown out. The question is how much work is required to do the throwing :)

**Comment by bennahugo** [comment]
You can just have a look and merge if you are happy with my changes?

**Comment by twillis449** [comment]
Ok - I hit the merge button :)

**Comment by bennahugo** [comment]
Great thanks. I'm busy reading the code on and off to see where I need to
make changes to cover the various tickets. I will try to build a module for
AIPS magnetosphere and add lunar ephemeris as first stabs I think

**Pull request merged**